Add kubectl logs feature and metrics-server deployment complete guide (#2112)
kubeedge/kubeedge: docs/setup/keadm.md

      tunnelPort: 10004
    ```

    Open the YAML file in edgecore:
    ``` shell
    sudo nano /etc/kubeedge/config/edgecore.yaml
    ```
    Modify the file in the following part (`enable: true`), (`server: 192.168.0.193:10004`):
    ``` yaml  
    edgeStream:
      enable: true
      handshakeTimeout: 30
      readDeadline: 15
      server: 192.168.0.139:10004
      tlsTunnelCAFile: /etc/kubeedge/ca/rootCA.crt
      tlsTunnelCertFile: /etc/kubeedge/certs/server.crt
      tlsTunnelPrivateKeyFile: /etc/kubeedge/certs/server.key
      writeDeadline: 15
    ```  

6. Restart all the cloudcore and edgecore.

    ``` shell
    sudo su
    ```
    cloudCore:
    ``` shell
    pkill cloudcore
    nohup cloudcore > cloudcore.log 2>&1 &
    ```
    edgeCore:
    ``` shell
    systemctl restart edgecore.service
    docker ps
    ```
    | CONTAINER ID  | NAMES                                           |
    |---------------|-------------------------------------------------|
    | d40ad4007d6f9 | k8s_POD_weave-net-fr9vx_kube-system_...         |
    | f59bdacbfc48  | k8s_kube-proxy_kube-proxy-s8476_kube-system_... |
    | f15a4b9aae7a  | k8s_POD_kube-proxy-s8476_kube-system_.....      |

>     Kill those `Network` related containers to make restart 

Author: @GsssC 
May I ask where should I put those settings in? Is it in edgecore.yaml?
Thank you very much in advance.

>     Kill those `Network` related containers to make restart 

Reviewer: @s099064146，those setting about `edgecoreService` is writen in  `/etc/kubeedge/edgecore.service`, i think note and let user modify it manually is enough.

    ```
   
    Go to the metrics server directory:

    ```bash
    cd metrics-server
    ```

    Make the docker image:

    ```bash
    make container
    ```

    Check if you have this docker image:

    ```bash
    docker images
    ```

    |                  REPOSITORY                           |                    TAG                   |   IMAGE ID   |     CREATE     |  SIZE  |
    |-------------------------------------------------------|------------------------------------------|--------------|----------------|--------|
    | gcr.io/k8s-staging-metrics-serer/ metrics-serer-amd64 | 6d92704c5a68cd29a7a81bce68e6c2230c7a6912 | a24f71249d69 | 19 seconds ago | 57.2MB |
    | metrics-server-kubeedge                               |                 latest                   | aef0fa7a834c | 28 seconds ago | 57.2MB |

    Make sure you change the tag of image by using its IMAGE ID to be compactable with image name in yaml file.

    ```bash
    docker tag a24f71249d69 metrics-server-kubeedge:latest
    ```

3. Apply the deployment yaml. For specific deployment documents, you can refer to https://github.com/kubernetes-sigs/metrics-server/tree/master/manifests.

>     Before you deploy metrics-server, you have to make sure that you deploy it on the node which has apiserver deployed on. In this case, that is the master node. As a consequence, it is needed to make master node schedulable by the following command:

Reviewer: The request to retrive metric-data from edgecore:10250 through tunnel between cloudcore and edgecore is issued by metric-server. To direct this req to cloudcore tunnel/stream server, we need iptables, so the importance is ** those iptables above must be applyed on the machine metric-server runs on**.
```
 iptables -t nat -A OUTPUT -p tcp --dport 10350 -j DNAT --to $CLOUDCOREIPS:10003
```
Of course, deploy ms on master is a good solution, because master has been applied those iptables for enabling logs feature , I think it is better to note the key point there as well.

    ``` shell
    kubectl taint nodes --all node-role.kubernetes.io/master-
    ```
   
    Then, in the deployment.yaml file, it must be specified that metrics-server is deployed on master node.
    (The hostname is chosen as the marked label.)
    In **metrics-server-deployment.yaml**
    ``` yaml
        spec:
          affinity:
            nodeAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                - matchExpressions:
                  #Specify which label in [kubectl get nodes --show-labels] you want to match
                  - key: kubernetes.io/hostname
                    operator: In
                    values:
                    #Specify the value in key
                    - charlie-latest
    ```

**IMPORTANT NOTE:**
1. Metrics-server needs to use hostnetwork network mode.

2. Use the image compiled by yourself and set imagePullPolicy to Never.

3. Enable the feature of --kubelet-use-node-status-port for Metrics-server

    Those settings need to be written in deployment yaml (metrics-server-deployment.yaml) file like this:

>     ``` yaml

Reviewer 2: maybe we can offer an entire deployment yaml?

Author: I did. But @Poor12 said that the original one is good enough yesterday, the one I used is here.

Ref:
 The deployment YAML files can be found in this repository:
    [DockerAndKubernetesCourseCode](https://github.com/DanWahlin/DockerAndKubernetesCourseCode/tree/master/samples/prometheus/metrics-server)

![image](https://user-images.githubusercontent.com/40239192/91865200-b80e7100-eca3-11ea-9ed2-770427391e82.png)

Reviewer 3: @luogangyi I have discussed this issue with fisher and daixiang. For kubeedge users, ms is not necessary and there is no need to write yaml files directly in kubeedge.

          volumes:
          # mount in tmp so we can safely use from-scratch images and/or read-only containers
          - name: tmp-dir
            emptyDir: {}
          hostNetwork: true                          #Add this line to enable hostnetwork mode
          containers:
          - name: metrics-server
            image: metrics-server-kubeedge:latest    #Make sure that the REPOSITORY and TAG are correct
            # Modified args to include --kubelet-insecure-tls for Docker Desktop (don't use this flag with a real k8s cluster!!)
            imagePullPolicy: Never                   #Make sure that the delpoyment uses the image you built up
            args:
              - --cert-dir=/tmp
              - --secure-port=4443
              - --v=2
              - --kubelet-insecure-tls

>               - --kubelet-preferred-address-types=InternalDNS,InternalIP,ExternalIP,Hostname

Reviewer 2: r u sure we should use InternalDNS here?

Author: I am not sure, but actually, I used this deployment file and successfully make it work.

      server: 192.168.0.139:10004
      tlsTunnelCAFile: /etc/kubeedge/ca/rootCA.crt
      tlsTunnelCertFile: /etc/kubeedge/certs/server.crt
      tlsTunnelPrivateKeyFile: /etc/kubeedge/certs/server.key
      writeDeadline: 15
    ```  

6. Restart all the cloudcore and edgecore.

    ``` shell
    sudo su
    ```
    cloudCore:
    ``` shell
    pkill cloudcore
    nohup cloudcore > cloudcore.log 2>&1 &
    ```
    edgeCore:
    ``` shell
    systemctl restart edgecore.service
    ```
    If you fail to restart edgecore, check if that is because of `kubeproxy` and kill it.  **kubeedge** reject it by default, we use a succedaneum called [edgemesh](https://github.com/kubeedge/kubeedge/blob/master/docs/proposals/edgemesh-design.md)

    **Note:** the importance is to avoid `kubeproxy` being deployed on edgenode. There are two methods to solve it:

    1. Add the following settings to `kubeproxy daemonset yaml`
    ``` yaml
    affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
              - matchExpressions:
                  - key: node-role.kubernetes.io/edge
                    operator: DoesNotExist
    ```

>     2. If you still want to run `kubeproxy`, ask **edgecore** not to check the environment by adding the setting in

Reviewer: ```suggestion
    2. If you still want to run `kubeproxy`, ask **edgecore** not to check the environment by adding the env variable in `edgecore.service` :
```

        ``` shell

>         sudo nano /etc/kubeedge/edgecore.service

Reviewer: centos no `nano`
```suggestion
        sudo vi /etc/kubeedge/edgecore.service
```

      tunnelPort: 10004
    ```

    Open the YAML file in edgecore:
    ``` shell
    sudo nano /etc/kubeedge/config/edgecore.yaml
    ```
    Modify the file in the following part (`enable: true`), (`server: 192.168.0.193:10004`):
    ``` yaml  
    edgeStream:
      enable: true
      handshakeTimeout: 30
      readDeadline: 15
      server: 192.168.0.139:10004
      tlsTunnelCAFile: /etc/kubeedge/ca/rootCA.crt
      tlsTunnelCertFile: /etc/kubeedge/certs/server.crt
      tlsTunnelPrivateKeyFile: /etc/kubeedge/certs/server.key
      writeDeadline: 15
    ```  

6. Restart all the cloudcore and edgecore.

    ``` shell
    sudo su
    ```
    cloudCore:
    ``` shell
    pkill cloudcore
    nohup cloudcore > cloudcore.log 2>&1 &
    ```
    edgeCore:
    ``` shell
    systemctl restart edgecore.service
    docker ps
    ```
    | CONTAINER ID  | NAMES                                           |
    |---------------|-------------------------------------------------|
    | d40ad4007d6f9 | k8s_POD_weave-net-fr9vx_kube-system_...         |
    | f59bdacbfc48  | k8s_kube-proxy_kube-proxy-s8476_kube-system_... |
    | f15a4b9aae7a  | k8s_POD_kube-proxy-s8476_kube-system_.....      |

>     Kill those `Network` related containers to make restart 

Reviewer 4: Do we need to restart all thses containers?

Author: @fisherxu 
In the general case, it is essentially important, unless the user has made some modification of their **edgecore.service** file or **kubeproxy daemonset yaml** file.

Reviewer 4: So I'm not sure why need to restart these proxy and cni releated containers?

Author: @fisherxu 
If I don't kill those containers first, the restart of edgecore will fail.

Reviewer 4: Thanks for the clarification, but I don't think we need to restart these containers.

If users want to upgrade or restart the edgecore, and the users apps must not be restart.

> If I don't kill those containers first, the restart of edgecore will fail.

What's the relationship between those containers and edgecore?

    ``` shell
    docker container kill d40ad407d6f9
    docker container kill f59bdacbfc48
    docker container kill f15a4b9aae7a
    systemctl restart edgecore.service
    ```

    **Note:** the importance is to avoid `kubeproxy` being deployed on edgenode. There are two methods to prevent it from happening
    1. Add the following settings to `kubeproxy daemonset yaml`
    ``` yaml
    affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
              - matchExpressions:
                  - key: node-role.kubernetes.io/edge
                    operator: DoesNotExist
    ```

>     2. Ask **edgecore** not to check the environment by adding the setting in

Reviewer: ```suggestion
    2. If you still want to run `kubeproxy`, ask **edgecore** not to check the environment by adding the setting in
```

```

### Enable `kubectl logs` Feature

> #### 1. Environments:

Reviewer 3: No need to write a specific environment.

Author: Okay, I will remove it.

    ```

3. Apply the deployment yaml. For specific deployment documents, you can refer to https://github.com/kubernetes-sigs/metrics-server/tree/master/manifests.

    Before you deploy metrics-server, you have to make sure that you deploy it on the node which has apiserver deployed on. In my case, that is my master node. As a consequence, I need to make master node schedulable by the following command:

    ``` shell
    kubectl taint nodes --all node-role.kubernetes.io/master-
    ```
   
    Then, in the deployment.yaml file, I must specify that metrics-server is deployed on master node.
    (I used hostname as the marked label.)
    In `metrics-server-deployment.yaml`
    ``` yaml
        spec:
          affinity:
            nodeAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                - matchExpressions:
                  #Specify which label in [kubectl get nodes --show-labels] you want to match
                  - key: kubernetes.io/hostname
                    operator: In
                    values:
                    #Specify the value in key
                    - charlie-latest
    ```

>     The deployment YAML files can be found in this repository:

Reviewer 3: There is no need for a specific yaml file, just the original way.

Author: Oh, I will remove it.
(Since I am a newbie and have no idea how to create it in an original way, I thought it is easier to have an example.)

    ``` yaml
          volumes:
          # mount in tmp so we can safely use from-scratch images and/or read-only containers
          - name: tmp-dir
            emptyDir: {}
          hostNetwork: true                          #Add this line to enable hostnetwork mode
          containers:
          - name: metrics-server
            image: metrics-server-kubeedge:latest    #Make sure that the REPOSITORY and TAG are correct
            # Modified args to include --kubelet-insecure-tls for Docker Desktop (don't use this flag with a real k8s cluster!!)
            imagePullPolicy: Never                   #Make sure that the delpoyment uses the image you built up
            args:
              - --cert-dir=/tmp
              - --secure-port=4443
              - --v=2
              - --kubelet-insecure-tls
              - --kubelet-preferred-address-types=InternalDNS,InternalIP,ExternalIP,Hostname
              - --kubelet-use-node-status-port       #Enable the feature of --kubelet-use-node-status-port for Metrics-server
            ports:
            - name: main-port
              containerPort: 4443
              protocol: TCP
    ```

    Metrics server works finally!

>     <img src="../images/metrics-server/final_result.png" width="713" height="376">

Reviewer 3: The pic is redundant.

Author: I see. I will delete the picture. Thanks!
